cpp/ps0006/ps0006_sol.cpp

```diff
@@ -4,43 +4,96 @@ using namespace std;
 
 bool input(int start, int end, int step) {
   if (step == 0) {
-    cout << "Error: Step cannot be 0." << endl;
+    if (start == end) // added a valid case
+      cout << start << endl;
+    else
+      cout << "Error: Step cannot be 0." << endl;
     return false;
   }
-  if ((step > 0 && start > end) || (step < 0 && start < end)) {
+  if ((start == end) || (step > 0 && start > end) || (step < 0 && start < end)) { // added a case for start == end
     cout << "Error: Start value cannot reach end value with the given step." << endl;
     return false;
   }
   return true;
 }
 
 
-void counter(int start = 0, int end = 0, int step = 1) {
+void counter(int end, int start = 0, int step = 1) { // swapped start with end
   if (!input(start, end, step)) {
     return;
   }
 
-  cout << "Counting from " << start << " to " << end << " with step " << step << ":" << endl;
 
-  for (int i = start; (step > 0 && i <= end) || (step < 0 && i >= end); i += step) {
-    cout << i << " ";
+  if ((start - end) % step == 0) // added a condition for invalid steps
+  {
+    cout << "Counting from " << start << " to " << end << " with step " << step << ":" << endl;
+    for (int i = start; (step > 0 && i <= end) || (step < 0 && i >= end); i += step) {
+      cout << i << " ";
+    }
+    cout << endl;
+  }
+  else
+  {
+    cout << "Error: Start value cannot reach end value with the given step." << endl;
   }
-  cout << endl;
 }
 
 int main() {
-  int start, end, step;
-
-  cout << "Enter start value (default 0): ";
-  cin >> start;
-
-  cout << "Enter end value: ";
-  cin >> end;
-
-  cout << "Enter step value (default 1): ";
-  cin >> step;
-
-  counter(start, end, step);
-
-  return 0;
+  // running my valid test cases 3/3
+    
+    counter(5); // 0 1 2 3 4 5
+    cout << "===============" << endl;
+    
+    counter(3, 1); // 1 2 3
+    cout << "===============" << endl;
+    
+    counter(-21, 6, -3); // 6 3 0 -3 -6 -9 -12 -15 -18 -21
+    cout << "===============" << endl;
+    
+    counter(10, 10, 0); // 10
+    cout << "===============" << endl;

```

Comment:
- [x] `start=0`, `step=1` as default
- [x] up counter works properly when `step` is positive
- [x] down counter works properly when `step` is negative